drv/cosmo-seq-server/src/main.rs

```diff
@@ -0,0 +1,595 @@
+                        _ => (),
+                    }
+                    hl::sleep_for(10);
+                }
+                if !okay {
+                    self.log_state_registers();
+                    self.log_pg_registers();
+                    // We can't do much else here, since we already cleared the
+                    // a0_en flag to disable the sequencer.
+                }
+
+                // Flip the host flash mux so the SP can read from it
+                // (this is secretly infallible on Cosmo, so we can unwrap it)
+                self.hf.set_mux(drv_hf_api::HfMuxState::SP).unwrap();
+            }
+
+            // This is purely an accounting change
+            (PowerState::A0, PowerState::A0PlusHP) => (),
+
+            _ => return Err(CpuSeqError::IllegalTransition),
+        }
+
+        self.state = state;
+        self.jefe.set_state(state as u32);
+        self.poke_timer();
+        Ok(())
+    }
+
+    /// Returns the current timer interval, in milliseconds
+    ///
+    /// If we are in `A0`, then we are waiting for the NIC to come up; if we are
+    /// in `A0PlusHP`, we're polling for a thermtrip or for someone disabling
+    /// the NIC.  In other states, there's no need to poll.
+    fn poll_interval(&self) -> Option<u32> {
+        match self.state {
+            PowerState::A0 => Some(10),
+            PowerState::A0PlusHP => Some(100),
+            _ => None,
+        }
+    }
+
+    /// Updates the system timer
+    fn poke_timer(&self) {
+        if let Some(interval) = self.poll_interval() {
+            set_timer_relative(interval, notifications::TIMER_MASK);
+        }
+    }
+}
+
+impl idl::InOrderSequencerImpl for ServerImpl {
+    fn get_state(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<PowerState, RequestError<core::convert::Infallible>> {
+        Ok(self.get_state_impl())
+    }
+
+    fn set_state(
+        &mut self,
+        _: &RecvMessage,
+        state: PowerState,
+    ) -> Result<(), RequestError<CpuSeqError>> {
+        self.set_state_impl(state, StateChangeReason::Other)?;
+        Ok(())
+    }
+
+    fn set_state_with_reason(
+        &mut self,
+        _: &RecvMessage,
+        state: PowerState,
+        reason: StateChangeReason,
+    ) -> Result<(), RequestError<CpuSeqError>> {
+        self.set_state_impl(state, reason)?;
+        Ok(())
+    }
+
+    fn send_hardware_nmi(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<(), RequestError<core::convert::Infallible>> {
+        // The required length for an NMI pulse is apparently not documented.
+        //
+        // Let's try 25 ms!
+        self.sys.gpio_reset(SP_TO_SP5_NMI_SYNC_FLOOD_L);
+        hl::sleep_for(25);
+        self.sys.gpio_set(SP_TO_SP5_NMI_SYNC_FLOOD_L);
+        Ok(())
+    }
+
+    fn read_fpga_regs(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<[u8; 64], RequestError<core::convert::Infallible>> {
+        // XXX todo
+        Ok([0; 64])
```

Comment:
Still TODO?

Reply:
This is not a meaningful API for `cosmo-seq`, so I'm going to return `ClientError::BadMessageContents` instead (which faults the caller).